### PR TITLE
Add `file_exists?` macro method

### LINKED ISSUE #10493: Add `#file_exists?` Top-level Macro Method
## Feature Request

I have a macro that checks for the existence of various template files, which may or may not exist, and renders the first template it finds. Imagine the following:

```crystal
ok "objects/show"
```

The macro "ok" expands to code like the following:

```crystal
if accepts?("application/activity+json", "application/json")
  {% if read_file?("#{basedir.id}/#{message.id}.json.ecr") %}
    halt env, status_code: {{code}}, response: render {{"#{basedir.id}/#{message.id}.json.ecr"}}
  {% end %}
end
if accepts?("text/html")
  {% if read_file?("#{basedir.id}/#{message.id}.html.ecr") %}
    halt env, status_code: {{code}}, response: render {{"#{basedir.id}/#{message.id}.html.ecr"}}, "src/views/layouts/default.html.ecr"
  {% elsif read_file?("#{basedir.id}/#{message.id}.html.slang") %}
    halt env, status_code: {{code}}, response: render {{"#{basedir.id}/#{message.id}.html.slang"}}, "src/views/layouts/default.html.ecr"
  {% end %}
end
if accepts?("text/plain")
  {% if read_file?("#{basedir.id}/#{message.id}.text.ecr") %}
...
```

Ignoring the omitted details of this specific use case, the issue is that, for a given controller action, the template files may not exist, and processing should fall through to more general purpose code later in the macro. Right now, the only way to check for the existence of a file is by trying to read the file.

I'd prefer an explicit check. Something like `exists?` or `file_exists?`. I've created a possible implementation here: https://github.com/crystal-lang/crystal/pull/10489

This should not pose compatibility problems with any existing code because there is currently no means to create new top-level functions available to macros outside of the compiler.

### PR DESCRIPTION
**Pull request by Sija**
Resolves #10493

### CODE REVIEW

**Comment by straight-shoota** [comment]
What's wrong with #10489?

**Comment by Sija** [comment]
@straight-shoota Nothing, I forgot about its existence 😅

**Comment by toddsundsted** [comment]
@Blacksmoke16 @Sija i don't think either of these were ever merged. i blew away the branch underlying my PR a short time ago, so my PR was closed. i'd still like to see this change, however.

**Comment by Sija** [comment]
@toddsundsted Sure, I reopened this PR as suggested + it's rebased and ready for review.
/cc @straight-shoota

File: src/compiler/crystal/macros.cr [hunk 1]
```diff
@@ -87,6 +87,10 @@ module Crystal::Macros
   def raise(message) : NoReturn
   end
 
+  # Returns `true` if the given *filename* exists, `false` otherwise.
+  def file_exists?(filename) : BoolLiteral
```
Comment (straight-shoota) [inline src/compiler/crystal/macros.cr:91]:
```suggestion
  def file_exists?(path) : BoolLiteral
```

Maybe `path` would be a better parameter name?

**Review by straight-shoota — APPROVED** [review]

File: src/compiler/crystal/macros.cr (hunk 1 above)
Comment (Sija) [reply src/compiler/crystal/macros.cr:91]: I'd agree, although it's the same as in `read_file` macro, so 🤷‍♂️

**Comment by Sija** [comment]
ping 🏓

**Review by beta-ziliani — APPROVED** [review]
Thanks @Sija !

**Review by straight-shoota — APPROVED** [review]

**Pull request merged**